torch/csrc/jit/codegen/cuda/executor.cpp

```diff
@@ -685,51 +747,105 @@ std::vector<at::Tensor> FusionExecutor::allocOutputs(
       auto output = kernel->outputs()[out_i]->as<TensorView>();
       if (alias_indices.count(out_i) != 0) {
         // aliasing to inputs, no need to allocate real output
-        outputs.push_back(
-            inferAndAlloc(output, {}, expr_eval, {}, options_, false));
+        // but we still need to push an entry here.
+        ret.push(int64_t(0));
       } else {
-        // Allocate a real output
-        outputs.push_back(
-            inferAndAllocOutput(output, expr_eval, options_, false));
+        // TODO: we are using meta here, which is bad since it doesn't account for devices. Switch to fake tensor instead
+        ret.push(inferAndAllocOutput(output, expr_eval, meta_options, false));
       }
     }
   }
-  return outputs;
+  return ret;
 }
 
-void FusionExecutor::setUsedTVs() {
-  auto used_vals = fusion_->usedMathVals();
-  auto used_tvs = ir_utils::filterByType<TensorView>(used_vals);
-  used_tvs_.clear();
+KernelArgumentHolder FusionExecutor::inferOutputSizes(
+    const KernelArgumentHolder& args,
+    const LaunchParams& launch_constraints) {
+  FUSER_PERF_SCOPE("FusionExecutor::RunFusion");
 
-  for (auto tv : used_tvs)
-    used_tvs_.push_back(tv);
+  ExecutorEntry* executor_entry = nullptr;
+  c10::optional<size_t> opt_code = args.getCacheId();
+  if (opt_code.has_value()) {
+    executor_entry = &executor_entry_lookup_[*opt_code];
+  }
+
+  executor_utils::initializeCudaContext();
+  TORCH_INTERNAL_ASSERT(lowered_);
+
+  TORCH_INTERNAL_ASSERT(
+      !executor_entry || !executor_entry->init,
+      "compile kernel shouldn't hit a pre-existing cache");
+  FUSER_PERF_SCOPE("ExecutorRunFusion::ValidateAndInitialize");
+  // TODO: validate kernel inputs currently won't be happy, since our fusion args are mapped with `meta` tensor instead of `cuda` tensor, check if this would be resolved with FakeTensor
+  // executor_utils::validateKernelInputs(fusion_, args, options_.device);
```

Comment:
Oh yes we are still validating inputs when we actually run the kernel.

The only issue here is that we use `meta` tensor right now to avoid allocating real memory for compilation. And that doesn't play well when we check devices on inputs. I tend to switch to fake tensor anyway, which would resolve the issue. So just temporarily commented this out.

Reply:
If we run this validation before launch then do we still need to validate here?

Reply:
I think it's still a good-to-have thing just to ensure that our output shape inference is doing the right thing.